packages/crypto/src/crypto.ts

```diff
@@ -23,3 +23,45 @@ export const hashLeftRight = (input1: any, input2: any) =>
 export const hashOne = (input: any) => hash1([input])
 
 export { SNARK_FIELD_SIZE, genRandomSalt, stringifyBigInts, unstringifyBigInts }
+
+export const genEpochKey = (
+    identityNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: bigint | number,
+    nonce: bigint | number,
+    maxEpochKey: bigint | number
+): bigint => {
+    const epochKey = hash4([
+        identityNullifier as any,
+        BigInt(attesterId),
+        epoch,
+        BigInt(nonce),
+    ]).valueOf()
+    // Adjust epoch key size according to epoch tree depth
+    const epochKeyModed = epochKey % BigInt(maxEpochKey)
+    return epochKeyModed
+}
+
+export const genEpochNullifier = (
+    identityNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: number | bigint
+): bigint => {
+    return hash3([BigInt(attesterId), BigInt(epoch), identityNullifier as any])
+}
+
+export const genStateTreeLeaf = (
+    idNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: bigint | number,
+    posRep: bigint | number,
+    negRep: bigint | number
+): BigInt => {
```

Comment:
I think we can either merge https://github.com/Unirep/Unirep/pull/140 first
Or directly change the type in the PR since there are many differences from #140

Reply:
Yea I think we can make divergent histories here. We put #140 on top of the v1.0.0 tag, and main is a separate history.

Reply:
f909e5b9611513963dbb65ce298e161e868bf56f